File: src/endpoints/tag/get_index.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token

async def get_index(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set data assets
        fields = []
        where = []
        values = {}
        join = []

        # set base params
        if params.name:
            where.append(f'AND name LIKE "%{params.name}%"')
        if params.module:
            fields = [ f'{Table.TAG.value}.*' ]
            join.append(f'JOIN {Table.MAP_TAG.value} ON {Table.MAP_TAG.value}.tag_srl = {Table.TAG.value}.srl')
            where.append(f'and module LIKE \'{params.module}\'')
        if params.module and params.module_srl:
            where.append(f'and module_srl = {params.module_srl}')

        # get total
        total = db.get_count(
            table_name=Table.TAG.value,
            column_name=f'DISTINCT {Table.TAG.value}.srl',
            where=where,
            join=join,
            values=values,
        )
        if not (total > 0): raise Exception('No data', 204)

        # get index
        index = db.get_items(
            table_name=Table.TAG.value,
            fields=fields,
            where=where,
            join=join,
            limit={ 'size': params.size, 'page': params.page },
            order={ 'order': params.order, 'sort': params.sort },
            values=values,
            unlimited=params.unlimited,
            duplicate=False,
        )
        print('???', total)

        # set result
        result = output.success({
            'message': 'Complete get tag index.',
            'data': {
                'total': total,
                'index': index,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/tag/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import Service from '@/classes/Service'
import { TagModel } from './__model'

type GetIndexParams = {
  query: TagModel['getIndexQuery']
  service: Service
}

export default async function getIndex({ query, service }: GetIndexParams)
{
  try
  {
    // set assets
    let _field = []
    let _where: string[] = []
    let _join: string[] = []
    let _values: ZZ = {}
    if (query.name)
    {
      _where.push(`AND t.name LIKE \'%${query.name}%\'`)
    }
    if (query.module)
    {
      _field.push(`t.*`)
      _join.push(`JOIN ${DB.TABLE.MAP_TAG} AS mt ON mt.tag_srl = t.srl`)
      _where.push(`AND mt.module LIKE \'${query.module}\'`)
    }
    if (query.module && query.module_srl)
    {
      _where.push(`AND mt.module_srl = ${query.module_srl}`)
    }

    // get count
    const count = db.getCount({
      table: `${DB.TABLE.TAG} AS t`,
      field: `COUNT(DISTINCT t.srl) AS count`,
      where: _where,
      join: _join,
      values: _values,
    })
    if (count.data <= 0) throw new ServiceError('No data', { status: 204 })

    // get index
    const index = db.getIndex({
      table: `${DB.TABLE.TAG} AS t`,
      prefix: 'DISTINCT',
      field: _field,
      where: _where,
      join: _join,
      order: query.order,
      page: query.page ?? 1,
      size: query.size ?? service.preference['tag.index.size'],
      values: _values,
    })

    return {
      total: count.data,
      index: index.data,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Tag index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
